core/src/main/scala/sttp/tapir/server/interceptor/metrics/MetricsEndpointInterceptor.scala

```diff
@@ -0,0 +1,115 @@
+    new EndpointHandler[F, B] {
+      override def onDecodeSuccess[I](
+          ctx: DecodeSuccessContext[F, I]
+      )(implicit monad: MonadError[F], bodyListener: BodyListener[F, B]): F[ServerResponse[B]] = {
+        if (ignoreEndpoints.contains(ctx.endpoint)) endpointHandler.onDecodeSuccess(ctx)
+        else {
+          val responseWithMetrics: F[ServerResponse[B]] = for {
+            _ <- collectMetrics { case EndpointMetric(Some(onRequest), _, _) => onRequest(ctx.endpoint) }
+            response <- endpointHandler.onDecodeSuccess(ctx)
+            withMetrics <- withBodyOnComplete(response) {
+              case Success(_) => collectMetrics { case EndpointMetric(_, Some(onResponse), _) => onResponse(ctx.endpoint, response) }
+              case _          => ().unit
+            }
+          } yield withMetrics
+
+          responseWithMetrics.handleError { case e: Exception =>
+            collectMetrics { case EndpointMetric(_, _, Some(onException)) => onException(ctx.endpoint, e) }.flatMap(_ => monad.error(e))
+          }
+        }
+      }
+
+      /** If there's some `ServerResponse` collects `onResponse` as well as `onRequest` metric which was not collected in `onDecodeSuccess` stage.
+        */
+      override def onDecodeFailure(
+          ctx: DecodeFailureContext
+      )(implicit monad: MonadError[F], bodyListener: BodyListener[F, B]): F[Option[ServerResponse[B]]] = {
+        if (ignoreEndpoints.contains(ctx.endpoint)) endpointHandler.onDecodeFailure(ctx)
+        else {
+          ctx.failure match {
+            case _: DecodeResult.Mismatch => endpointHandler.onDecodeFailure(ctx)
+            case _ =>
+              val responseWithMetrics: F[Option[ServerResponse[B]]] = for {
+                response <- endpointHandler.onDecodeFailure(ctx)
+                withMetrics <- response match {
+                  case Some(response) =>
+                    for {
+                      _ <- collectMetrics { case EndpointMetric(Some(onRequest), _, _) => onRequest(ctx.endpoint) }
+                      res <- withBodyOnComplete(response) {
+                        case Success(_) =>
+                          collectMetrics { case EndpointMetric(_, Some(onResponse), _) => onResponse(ctx.endpoint, response) }
+                        case _ => ().unit
+                      }
+                    } yield Some(res)
+                  case None => monad.unit(None)
+                }
+              } yield withMetrics
+
+              responseWithMetrics.handleError { case e: Exception =>
+                collectMetrics { case EndpointMetric(_, _, Some(onException)) => onException(ctx.endpoint, e) }.flatMap(_ => monad.error(e))
+              }
+          }
+        }
+      }
+    }
+
+  private def collectMetrics(pf: PartialFunction[EndpointMetric[F], F[Unit]])(implicit monad: MonadError[F]): F[Unit] = {
+    def collect(metrics: List[EndpointMetric[F]]): F[Unit] = {
```

Comment:
that's usually called `sequence` (a function `List[F[T]] => F[List[T]]` - which we kind of have here)